blog-posts/cdk/portfolio/article.md

```diff
@@ -0,0 +1,228 @@
+- Create a CloudFront distribution, allowing to distribute my website all around the world, and to use my new HTTPS certificate.
+- Create DNS A records in my hosted zone, to redirect traffic from `pchol.fr` and `www.pchol.fr` to the CloudFront distribution.
+
+Seems complicated ? These 4 steps can be coded in a few lines thanks to the CDK, everything left to do is to launch the deployment!
+
+Let's create the hosted zone, by adding this snippet of code under the S3 Bucket definition:
+
+```typescript
+const DOMAIN_NAME = 'pchol.fr'; // your domain name
+
+// Create a Route53 hosted zone to later create DNS records
+// ⚠️ Manual action required: when the hosted zone was created, copy its NS records into your domain's name servers
+const hostedZone = new HostedZone(this, 'DomainHostedZone', {
+  zoneName: DOMAIN_NAME,
+});
+```
+
+Then deploy this change to create the hosted zone.
+
+```sh
+cdk deploy
+```
+
+I have to perform **the only manual input required in this tutorial (⚠️ very important ⚠️)**: after deployment, Route 53 will create a `NS` record containing 4 values in my new hosted zone, I have to copy them into my DNS's name servers (in order to prove my ownership of the domain).
+
+_NS records in the new hosted zone_ ![NS records](./assets/records.png 'NS records')
+
+_Copied into the domain name servers_ ![Name servers](./assets/name-servers.png 'Name servers')
+
+On the screenshots, I did it using a domain name bought on AWS, but it can also be achieved on other providers like [gandi][gandi].
+
+Now that I'm done with the domain, I can create the certificate, and set the validation method to `fromDns`. It will automatically communicate with my freshly created hosted zone to validate the authenticity of the certificate.
+
+_⚠️ The certificate MUST be created in the region us-east-1 to be compatible with CloudFront, that's why I deploy all my resources in this region. CloudFront allows for "at edge" delivery, which means that it's not a problem if your visitors are not in the US._
+
+```typescript
+const WWW_DOMAIN_NAME = `www.${DOMAIN_NAME}`;
+
+// Create the HTTPS certificate (⚠️ must be in region us-east-1 ⚠️)
+const httpsCertificate = new Certificate(this, 'HttpsCertificate', {
+  domainName: DOMAIN_NAME,
+  subjectAlternativeNames: [WWW_DOMAIN_NAME],
+  validation: CertificateValidation.fromDns(hostedZone),
+});
+```
+
+Then, time for the CloudFront distribution. It will allow me to distribute the content of the bucket to my end users all around the world in record time using caching. I specify my bucket as the origin, enable a REDIRECT_TO_HTTPS policy, set the domain names (with the addition of the www subdomain for optimal compatibility with all browsers), and finally reference my new https certificate.
+
+```typescript
+// Create the CloudFront distribution linked to the website hosting bucket and the HTTPS certificate
+const cloudFrontDistribution = new Distribution(this, 'CloudFrontDistribution', {
+  defaultBehavior: {
```

Comment:
Ça vaut pe le cout d'ajouter la security headers policy. Est-ce que ton déploiement est compliant sls-mentor ?

Reply:
I also specified an optional responseHeadersPolicy, with the ID corresponding to the AWS `Managed-SecurityHeadersPolicy`. It's a quick win to improve security in my distribution. (This best practice is part of [sls-mentor][sls-mentor], feel free to check it out!)

```typescript
// Create the CloudFront distribution linked to the website hosting bucket and the HTTPS certificate
const cloudFrontDistribution = new Distribution(this, 'CloudFrontDistribution', {
  defaultBehavior: {
    origin: new S3Origin(staticWebsiteHostingBucket, {}),
    viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
    responseHeadersPolicy: { responseHeadersPolicyId: '67f7725c-6f97-4210-82d7-5512b31e9d03' },
  },
  domainNames: [DOMAIN_NAME, WWW_DOMAIN_NAME],
  certificate: httpsCertificate,
});
```

Reply:
en effet !